cudamapper/src/matcher_gpu.cu

```diff
@@ -0,0 +1,234 @@
+/*
+* Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#include "matcher_gpu.cuh"
+
+#include <thrust/scan.h>
+#include <thrust/transform_scan.h>
+#include <thrust/execution_policy.h>
+
+#include <claragenomics/utils/cudautils.hpp>
+#include <claragenomics/utils/mathutils.hpp>
+#include <claragenomics/utils/signed_integer_utils.hpp>
+
+namespace
+{
+template <typename RandomAccessIterator, typename ValueType>
+__device__ RandomAccessIterator lower_bound(RandomAccessIterator lower_bound, RandomAccessIterator upper_bound, ValueType query)
+{
+    while (upper_bound - lower_bound > 0)
```

Comment:
This shouldn't be a problem for valid input, because upper_bound - lower_bound will never be less than 0, but the `upper_bound > lower_bound` test will give defined behavior for invalid input as well. I also added an assert to check for valid input.